src/Parsing/Impl/Ast/ScopeStatement.cs

```diff
@@ -15,274 +15,47 @@
-        private List<string> _globalVars;                               // global variables accessed from this scope
-        private List<string> _cellVars;                                 // variables accessed from nested scopes
-        private List<NameExpression> _nonLocalVars;                             // variables declared as nonlocal within this scope
-        private Dictionary<string, List<PythonReference>> _references;        // names of all variables referenced, null after binding completes
-
-        internal const string NameForExec = "module: <exec>";
-
-        public ScopeStatement Parent { get; set; }
-
-        public abstract Statement Body {
-            get;
+        public ScopeStatement() {
+            _body = new EmptyStatement();
         }
 
-        internal bool ContainsImportStar { get; set; }
-        internal bool ContainsExceptionHandling { get; set; }
-
-        internal bool ContainsUnqualifiedExec { get; set; }
-
-        /// <summary>
-        /// True if this scope accesses a variable from an outer scope.
-        /// </summary>
-        public bool IsClosure => FreeVariables != null && FreeVariables.Count > 0;
-
-        /// <summary>
-        /// True if an inner scope is accessing a variable defined in this scope.
-        /// </summary>
-        public bool ContainsNestedFreeVariables { get; set; }
-
-        /// <summary>
-        /// True if we are forcing the creation of a dictionary for storing locals.
-        /// 
-        /// This occurs for calls to locals(), dir(), vars(), unqualified exec, and
-        /// from ... import *.
-        /// </summary>
-        public bool NeedsLocalsDictionary { get; set; }
-
-        public virtual string/*!*/ Name => "<unknown>";
-
-        /// <summary>
-        /// True if variables can be set in a late bound fashion that we don't
-        /// know about at code gen time - for example via from fob import *.
-        /// 
-        /// This is tracked independently of the ContainsUnqualifiedExec/NeedsLocalsDictionary
-        /// </summary>
-        internal virtual bool HasLateBoundVariableSets { get; set; }
-
-        internal Dictionary<string, PythonVariable> Variables { get; private set; }
-
         /// <summary>
-        /// Gets the variables for this scope.
+        /// For backwards compatibility, the closest Scope statement this scope statement was declared in
         /// </summary>
-        public ICollection<PythonVariable> ScopeVariables
-            => Variables?.Values ?? Array.Empty<PythonVariable>() as ICollection<PythonVariable>;
-
-        public virtual bool IsGlobal => false;
-
-        public PythonAst GlobalParent {
-            get {
-                var cur = this;
-                while (!(cur is PythonAst)) {
-                    Debug.Assert(cur != null);
-                    cur = cur.Parent;
-                }
-                return (PythonAst)cur;
-            }
-        }
-
-        protected void Clear() {
```

Comment:
Similarly protected here, so you'd need to preserve this.

Reply:
Done